libs/rapidxml-1.13/rapidxml_print.hpp

```diff
@@ -101,6 +101,35 @@ namespace rapidxml
 
         ///////////////////////////////////////////////////////////////////////////
         // Internal printing operations
+
+	// Doing template definitions first, due to GCC name lookup changes, see:
+	// https://stackoverflow.com/questions/14113923/rapidxml-print-header-has-undefined-methods
+	template<class OutIt, class Ch>
+	inline OutIt print_children(OutIt out, const xml_node<Ch> *node, int flags, int indent);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_attributes(OutIt out, const xml_node<Ch> *node, int flags);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_data_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_cdata_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_element_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_declaration_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_comment_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_doctype_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);
+
+	template<class OutIt, class Ch>
+	inline OutIt print_pi_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);
```

Comment:
Oh, that's a nice workaround!
Might be a good idea to add a guard for gcc only, since this only happens on gcc 4.7 or newer, if I understood correctly:

```suggestion

#ifdef __GNUC__
	#if __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 7)

	// Doing template definitions first, due to GCC name lookup changes, see:
	// https://stackoverflow.com/questions/14113923/rapidxml-print-header-has-undefined-methods
	template<class OutIt, class Ch>
	inline OutIt print_children(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	template<class OutIt, class Ch>
	inline OutIt print_attributes(OutIt out, const xml_node<Ch> *node, int flags);

	template<class OutIt, class Ch>
	inline OutIt print_data_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	template<class OutIt, class Ch>
	inline OutIt print_cdata_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	template<class OutIt, class Ch>
	inline OutIt print_element_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	template<class OutIt, class Ch>
	inline OutIt print_declaration_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	template<class OutIt, class Ch>
	inline OutIt print_comment_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	template<class OutIt, class Ch>
	inline OutIt print_doctype_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	template<class OutIt, class Ch>
	inline OutIt print_pi_node(OutIt out, const xml_node<Ch> *node, int flags, int indent);

	#endif
#endif

```

Reply:
Makes sense!

Reply:
Actually, now that I think about it, macos uses clang so this might break mac builds? Let me double check.

Reply:
After testing this does break builds on MacOS. In trying to find which version of clang would need to be checked for I ran into [this 9 year old stackoverflow question](https://stackoverflow.com/questions/7369314/gcc-clang-and-ibm-disagree-on-how-to-perform-template-parameter-dependent-name).

So I would suggest either a general check for clang using ``__clang__`` or (if I understand the page correctly, which goes a bit above my head) including it for all compilers, as it seems not defined by the standard to fail without the definitions first but it is the best interpretation of the standard (and thus leaving the code as is). My preference goes to the latter but up to you.